{% extends "home/base.html" %}
{% load static %}

{% block content %}
<div class="container py-4">

    <div class="p-3 mb-2 bg-light rounded-3">
        <div class="container-fluid py-3">
          <div class="row">
            <!-- Text Column -->
            <div class="col">
              <h2 class="display-6 fw-bold">Example case of ChemoPar-db usage</h2>
              <hr class="my-2">
                <p>
                    Here we provide an example of the human CXCL12 chemokine. See the original page <a style="text-decoration:none;" href="/protein/CXCL12_HUMAN/" target="_blank" class="">here</a>.
                </p>
            </div>  
          </div>
        </div>    
      </div>

    <h2 class="mt-3 mb-3 search-title">Protein information</h2>
    <p>
        The top section of the CXCL12 protein page gives an overview of general information including alternate names, organisms, and links to external databases (UniProt, Guide to Pharmacology) 
        as well as information in the corresponding UniProt record (ID, Organism, protein length).
    </p>
    <p>
        Then there is a representation of the canonical protein sequence of the specific entry. This horizontal scrollable table shows all amino acids, their generic numbers, and is divided into
        the different structural segments (N-term, CXC, N-loop, B-sheet 1, 30s-loop, B-sheet 2, 40s-loop, B-sheet 3, 50s-loop, a-helix, and C-term).
    </p>

    <img class="rounded shadow-twcss mx-auto my-4 d-block" src="{% static 'home/images/example_proteinpage.png' %}" alt="Protein page general information" style="border-radius=0.3rem; border:1px solid rgba(0,0,0,.1)" width="80%">

    <p>
        Below that, there is a chemokine diagram that contains a 2D depiction of the canonical residues with different sections named.
        This diagram is interactive and can be customized by the user. The code to generate this diagram is adapted from the GPCRdb<sup><a href="https://gpcrdb.org/" target="_blank">1</a></sup> snake plot and customized for chemokine representation.
    </p>

    <img class="rounded shadow-twcss mx-auto my-4 d-block" src="{% static 'home/images/example_diagram.png' %}" alt="" style="border-radius=0.3rem; border:1px solid rgba(0,0,0,.1)" width="50%">

    <p>
        If there are any structures of the specific protein in the database, they will be displayed. Clicking on any of these structures will redirect you to the detailed structure page for further information.
    </p>

    <img class="rounded shadow-twcss mx-auto my-4 d-block" src="{% static 'home/images/example_structure_list.png' %}" alt="" style="border-radius=0.3rem; border:1px solid rgba(0,0,0,.1)" width="50%">


    <h2 class="mt-3 mb-3 search-title">Structure information</h2>

    <p>
      The structure page provides comprehensive details about the structure, including links to the PDB website, UniProt, and related publication. Additionally, it offers a link back to the protein page of ChemoPar-db. 
      If any interactions are documented in the database, links to these interactions are provided, categorized by each chemokine chain in the structure.
    </p>

    <p>
      The page also features a chemokine sequence alignment, comparing the structure with the canonical UniProt sequence. This alignment is available for download in both .csv and .aln formats. Lastly, all entities present in the structure are listed for easy reference.
    </p>

    <img class="rounded shadow-twcss mx-auto my-4 d-block" src="{% static 'home/images/example_structure_page.png' %}" alt="" style="border-radius=0.3rem; border:1px solid rgba(0,0,0,.1)" width="50%">

    <h2 class="mt-3 mb-3 search-title">Interaction page</h2>

    <p>
      The Interaction page again provides detailed information about the structure, featuring the 3D viewer where the main chemokine chain is displayed in grey and any partner chains are highlighted in blue. 
      Users can select a partner in the chemokine diagram to highlight the interacting residues as mentioned in the publication. Additionally, a comprehensive list of all interactions is available for download in Excel format.
    </p>

    <img class="rounded shadow-twcss mx-auto my-4 d-block" src="{% static 'home/images/example_interaction_page.png' %}" alt="" style="border-radius=0.3rem; border:1px solid rgba(0,0,0,.1)" width="50%">

    <h2 class="mt-3 mb-3 search-title">Interaction Fingerprint Search</h2>

    <p>
      At the bottom of the Interaction Page, users can perform an Interaction Fingerprint (IFP) search for a selected chemokine-partner combination. This feature calculates the Tanimoto similarity between the IFP bit-string of the query and the entire database. Upon completion, users are directed to a results page that lists all structures along with their similarity scores.
    </p>

    <img class="rounded shadow-twcss mx-auto my-4 d-block" src="{% static 'home/images/example_ifpsearch.png' %}" alt="" style="border-radius=0.3rem; border:1px solid rgba(0,0,0,.1)" width="50%">


  </div>



{% endblock %}